iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -2172,6 +2178,80 @@ static STRING_HANDLE buildClientId(const char* device_id, const char* module_id)
     }
 }
 
+static int buildConfigForUsernameStep2IfNeeded(PMQTTTRANSPORT_HANDLE_DATA transport_data)
+{
+    int result;
+
+    if (!transport_data->isConnectUsernameSet)
+    {
+        STRING_HANDLE versionAndClientType = NULL;
+        STRING_HANDLE modelIdParameter = NULL;
+        STRING_HANDLE urlEncodedModelId = NULL;
+        const char* modelId = transport_data->transport_callbacks.get_model_id_cb(transport_data->transport_ctx);
+        // TODO: The preview API version in SDK is only scoped to scenarios that require the modelId to be set.
+        // https://github.com/Azure/azure-iot-sdk-c/issues/1547 tracks removing this once non-preview API versions support modelId.
+        const char* apiVersion = (modelId != NULL) ? IOTHUB_API_PREVIEW_VERSION : IOTHUB_API_VERSION;
+        const char* appSpecifiedProductInfo = transport_data->transport_callbacks.prod_info_cb(transport_data->transport_ctx);
+        STRING_HANDLE productInfoEncoded = NULL; 
+
+        if ((productInfoEncoded = URL_EncodeString((appSpecifiedProductInfo != NULL) ? appSpecifiedProductInfo : DEFAULT_IOTHUB_PRODUCT_IDENTIFIER)) == NULL)
+        {
+            LogError("Unable to UrlEncode productInfo");
+            result = MU_FAILURE;
+        }
+        else if ((versionAndClientType = STRING_construct_sprintf("?api-version=%s&DeviceClientType=%s", apiVersion, STRING_c_str(productInfoEncoded))) == NULL)
+        {
+            LogError("Failed constructing string");
+            result = 0;
+        }
+        else if (STRING_concat_with_STRING(transport_data->configPassedThroughUsername, versionAndClientType) != 0)
+        {
+            LogError("Failed concatenating the product info");
+            result = 0;
+        }           
+        else if (modelId != NULL)
+        {
+            if ((urlEncodedModelId = URL_EncodeString(modelId)) == NULL)
+            {
+                LogError("Failed to URL encode the modelID string");
+                result = MU_FAILURE;
+            }
+            else if ((modelIdParameter = STRING_construct_sprintf("&%s=%s", DT_MODEL_ID_TOKEN, STRING_c_str(urlEncodedModelId))) == NULL)
+            {
+                LogError("Cannot build modelID string");
+                result = MU_FAILURE;
+            }
+            else if (STRING_concat_with_STRING(transport_data->configPassedThroughUsername, modelIdParameter) != 0)
+            {
+                LogError("Failed to set modelID parameter in connect");
+                result = MU_FAILURE;
+            }
+            else
+            {
+                result = 0;
+            }
+        }
+        else
+        {
+            result = 0;
+        }
+
+        // setting connect string is only allowed once in the lifetime of the device client.
+        transport_data->isConnectUsernameSet = true;
```

Comment:
This is set to true even in the case that an above call failed. Expected behavior?

Reply:
Yup, I knew that one going in.  The issue why I initially had it like this is we're modifying `transport_data->configPassedThroughUsername` in middle of this function and where later stuff can fail.  This means we can't possibly guess its state if we come through the `step2` function again.

The fix is to do everything in temporary storage, and then only set `configPassedThroughUsername` if everything is successful (and only set `isConnectUsernameSet=true` after).  Even though it's massively unlikely we'd recover if we hit an out of memory at this stage, more I think about it the less I like this.  I'm going to fixup esp. as amount of extra memory isn't that large compared to TLS setup & we delete inside this function itself.